contractcourt/htlc_incoming_contest_resolver.go

```diff
@@ -440,18 +483,92 @@ func (h *htlcIncomingContestResolver) SupplementState(_ *channeldb.OpenChannel)
 }
 
 // decodePayload (re)decodes the hop payload of a received htlc.
-func (h *htlcIncomingContestResolver) decodePayload() (*hop.Payload, error) {
+func (h *htlcIncomingContestResolver) decodePayload() (*hop.Payload,
+	[]byte, error) {
+
 	onionReader := bytes.NewReader(h.htlc.OnionBlob)
 	iterator, err := h.OnionProcessor.ReconstructHopIterator(
 		onionReader, h.htlc.RHash[:],
 	)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
-	return iterator.HopPayload()
+	payload, err := iterator.HopPayload()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Transform onion blob for the next hop.
+	var onionBlob [lnwire.OnionPacketSize]byte
+	buf := bytes.NewBuffer(onionBlob[0:0])
+	err = iterator.EncodeNextHop(buf)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return payload, onionBlob[:], nil
 }
 
 // A compile time assertion to ensure htlcIncomingContestResolver meets the
 // ContractResolver interface.
 var _ htlcContractResolver = (*htlcIncomingContestResolver)(nil)
+
+// interceptedForward implements the on-chain behavior for the resolution of
+// a forwarded htlc.
+type interceptedForward struct {
+	packet      *InterceptedPacket
+	resolveChan chan lntypes.Preimage
+}
+
+func newInterceptedForward(
+	packet *InterceptedPacket,
+	resolveChan chan lntypes.Preimage) *interceptedForward {
+
+	return &interceptedForward{
+		resolveChan: resolveChan,
+		packet:      packet,
+	}
+}
+
+// Packet returns the intercepted htlc packet.
+func (f *interceptedForward) Packet() InterceptedPacket {
+	return *f.packet
+}
+
+// Resume notifies the intention to resume an existing hold forward. This
+// basically means the caller wants to resume with the default behavior for this
+// htlc which usually means forward it.
+func (f *interceptedForward) Resume() error {
+	return errors.New("cannot resume in the on-chain flow")
+}
+
+// Fail notifies the intention to fail an existing hold forward with an
+// encrypted failure reason.
+func (f *interceptedForward) Fail(_ []byte) error {
+	// We can't actively fail an htlc. The best we can do is abandon the
+	// resolver. This is not a safe operation though, because there may be a
+	// race with the preimage beacon supplying a preimage.
+	return errors.New("cannot fail in the on-chain flow")
```

Comment:
Couldn't we get around the preimage race with two way handoff? Get the resolver to close a `done` channel if we select the fail case or a `quit` channel if it's finished resolving (we can then error because we didn't get a signal from `done` first)

Reply:
Yes, maybe that can work. But I don't know if it is really worth the extra code. The chances of this race happening are already very small. So we'd only do this for the case where the race indeed happens and the external fail signal is received milliseconds before a preimage comes in from a different source. If it were the other way around, returning an error (which we always do in the current version of this pr) would be the correct behavior anyway.

The preimage coming in from another source itself already seems to be quite unlikely. Typically the preimage is learned after a successful forward. If the htlc is still held, there can not yet be a successful forward, unless there is another htlc in the system with the same hash. Mpp perhaps.

Given that it seems that the probability of this race actually occurring is really very low and also that in that case the current implementation returning an error is not incorrect, I don't think it is necessary to complicate the logic with a result channel. But I may miss something here. Definitely interested in hearing your opinion on it.